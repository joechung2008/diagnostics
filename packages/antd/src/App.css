html,
body {
  height: 100%;
  margin: 0;
  padding: 0;
}

*,
*::after,
*::before {
  box-sizing: border-box;
}

.centered {
  align-items: center;
  display: flex;
  height: 100%;
  justify-content: center;
  width: 100%;
}

.extension-root {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  max-height: max-content;
  overflow-y: auto;
}

.flexbox {
  display: flex;
  flex-direction: column;
  height: 100vh;
}

.grow {
  flex: 1;
}

.stack {
  display: flex;
  flex-direction: row;
  height: 100%;
}

.tab-list {
  padding-left: 1rem;
}

.tab-panel {
  flex: 1;
  overflow-y: auto;
}
