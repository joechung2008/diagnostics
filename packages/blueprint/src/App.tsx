import {
  Button,
  ButtonGroup,
  NonIdealState,
  Spinner,
  Tab,
  Tabs,
} from "@blueprintjs/core";
import clsx from "clsx";
import { useCallback, useMemo, useState } from "react";
import useSWR from "swr";
import BuildInfo from "./BuildInfo";
import Extension from "./Extension";
import Extensions from "./Extensions";
import ServerInfo from "./ServerInfo";
import { isExtensionInfo } from "./utils";

const Environment = {
  Public: "https://hosting.portal.azure.net/api/diagnostics",
  Fairfax: "https://hosting.azureportal.usgovcloudapi.net/api/diagnostics",
  Mooncake: "https://hosting.azureportal.chinacloudapi.cn/api/diagnostics",
} as const;

type EnvironmentUrl =
  | "https://hosting.portal.azure.net/api/diagnostics"
  | "https://hosting.azureportal.usgovcloudapi.net/api/diagnostics"
  | "https://hosting.azureportal.chinacloudapi.cn/api/diagnostics";

interface AppProps {
  className?: string;
}

const App = ({ className }: AppProps) => {
  const [environment, setEnvironment] = useState<EnvironmentUrl>(
    Environment.Public
  );
  const [extension, setExtension] = useState<ExtensionInfo>();
  const [selectedTab, setSelectedTab] = useState<string>("extensions");

  const {
    data: diagnostics,
    error,
    isLoading,
  } = useSWR(environment, async (environment: string) => {
    const response = await fetch(environment);
    if (!response.ok) {
      throw new Error(`Failed to fetch diagnostics: ${response.statusText}`);
    }

    return response.json();
  });

  const showPaasServerless = useMemo(
    () => isExtensionInfo(diagnostics?.extensions["paasserverless"]),
    [diagnostics?.extensions]
  );

  const handleLinkClick = useCallback(
    (item?: KeyedNavLink) => {
      if (item) {
        const extension = diagnostics?.extensions[item.key];
        if (isExtensionInfo(extension)) {
          setExtension(extension);
        }
      }
    },
    [diagnostics?.extensions]
  );

  return (
    <div className={clsx("flexbox", className)}>
      <ButtonGroup>
        <Button
          active={environment === Environment.Public}
          variant="minimal"
          onClick={() => {
            setEnvironment(Environment.Public);
            setExtension(undefined);
            setSelectedTab("extensions");
          }}
        >
          Public Cloud
        </Button>
        <Button
          active={environment === Environment.Fairfax}
          variant="minimal"
          onClick={() => {
            setEnvironment(Environment.Fairfax);
            setExtension(undefined);
            setSelectedTab("extensions");
          }}
        >
          Fairfax
        </Button>
        <Button
          active={environment === Environment.Mooncake}
          variant="minimal"
          onClick={() => {
            setEnvironment(Environment.Mooncake);
            setExtension(undefined);
            setSelectedTab("extensions");
          }}
        >
          Mooncake
        </Button>
        <span className="button-group-separator" />
        {showPaasServerless && (
          <Button
            variant="minimal"
            onClick={() => {
              const paasserverless = diagnostics?.extensions["paasserverless"];
              if (isExtensionInfo(paasserverless)) {
                setExtension(paasserverless);
                setSelectedTab("extensions");
              }
            }}
          >
            paasserverless
          </Button>
        )}
        <Button
          variant="minimal"
          onClick={() => {
            const websites = diagnostics?.extensions["websites"];
            if (isExtensionInfo(websites)) {
              setExtension(websites);
              setSelectedTab("extensions");
            }
          }}
        >
          websites
        </Button>
      </ButtonGroup>
      <Tabs
        selectedTabId={selectedTab}
        size="large"
        onChange={(tabId) => setSelectedTab(tabId as string)}
      >
        <Tab
          id="extensions"
          aria-controls="extensions-tab"
          title="Extensions"
        />
        <Tab id="build" aria-controls="build-tab" title="Build Information" />
        <Tab
          id="server"
          aria-controls="server-tab"
          title="Server Information"
        />
      </Tabs>
      {isLoading ? (
        <div className="tab-panel vertical-center">
          <Spinner aria-label="Loading..." />
        </div>
      ) : error ? (
        <NonIdealState icon="error" title="Error" description={error.message} />
      ) : selectedTab === "extensions" && diagnostics?.extensions ? (
        <div id="extensions-tab" className="tab-panel">
          <div className="stack">
            <Extensions
              extensions={diagnostics.extensions}
              onLinkClick={handleLinkClick}
            />
            {extension && <Extension {...extension} />}
          </div>
        </div>
      ) : selectedTab === "build" && diagnostics?.buildInfo ? (
        <div id="build-tab" className="tab-panel">
          <BuildInfo {...diagnostics.buildInfo} />
        </div>
      ) : selectedTab === "server" && diagnostics?.serverInfo ? (
        <div id="server-tab" className="tab-panel">
          <ServerInfo {...diagnostics.serverInfo} />
        </div>
      ) : null}
    </div>
  );
};

export default App;
