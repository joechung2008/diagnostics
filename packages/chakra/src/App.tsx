import BuildInfo from "@/BuildInfo";
import { ColorModeButton } from "@/components/ui/ColorMode";
import Extension from "@/Extension";
import Extensions from "@/Extensions";
import ServerInfo from "@/ServerInfo";
import { isExtensionInfo } from "@/utils";
import {
  Box,
  Button,
  Flex,
  Menu,
  Portal,
  Tabs,
  Alert,
  Spinner,
} from "@chakra-ui/react";
import {
  startTransition,
  useCallback,
  useMemo,
  useState,
  Activity,
} from "react";
import useSWR from "swr";

const Environment = {
  Public: "https://hosting.portal.azure.net/api/diagnostics",
  Fairfax: "https://hosting.azureportal.usgovcloudapi.net/api/diagnostics",
  Mooncake: "https://hosting.azureportal.chinacloudapi.cn/api/diagnostics",
} as const;

type EnvironmentUrl = (typeof Environment)[keyof typeof Environment];

function getEnvironnmentName(environment: EnvironmentUrl | undefined): string {
  switch (environment) {
    case Environment.Public:
      return "Public Cloud";
    case Environment.Fairfax:
      return "Fairfax";
    case Environment.Mooncake:
      return "Mooncake";
    default:
      return "Select environment";
  }
}

const App: React.FC = () => {
  const [environments, setEnvironments] = useState<
    Record<string, EnvironmentUrl[]>
  >({
    environment: [Environment.Public],
  });
  const environment = useMemo<EnvironmentUrl>(
    () => environments.environment[0],
    [environments.environment]
  );
  const {
    data: diagnostics,
    error,
    isLoading,
  } = useSWR(environment, async (url: string): Promise<Diagnostics> => {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`Failed to fetch diagnostics: ${response.statusText}`);
    }

    return response.json();
  });
  const [extension, setExtension] = useState<ExtensionInfo>();
  const [selectedTab, setSelectedTab] = useState<string>("extensions");

  const handleLinkClick: ExtensionsProps["onLinkClick"] = useCallback(
    (_, item) => {
      if (item) {
        const $extension = diagnostics?.extensions[item.key];
        if (isExtensionInfo($extension)) {
          setExtension($extension);
        }
      }
    },
    [diagnostics?.extensions]
  );

  const showPaasServerless = useMemo(
    () => isExtensionInfo(diagnostics?.extensions["paasserverless"]),
    [diagnostics?.extensions]
  );

  const environmentName = useMemo(
    () => getEnvironnmentName(environment),
    [environment]
  );

  const handleEnvironmentChange = useCallback((value: string) => {
    if (value) {
      startTransition(() => {
        setEnvironments((previous) => ({
          ...previous,
          environment: [value as EnvironmentUrl],
        }));
        setExtension(undefined);
        setSelectedTab("extensions");
      });
    }
  }, []);

  return (
    <Flex flexDirection="column" h="100vh" p="1">
      <Flex justify="space-between">
        <Flex gap="1">
          <Menu.Root>
            <Menu.Trigger asChild>
              <Button variant="outline">{environmentName}</Button>
            </Menu.Trigger>
            <Portal>
              <Menu.Positioner>
                <Menu.Content>
                  {Object.entries(Environment).map(([key, value]) => (
                    <Menu.Item
                      key={key}
                      value={value}
                      onClick={() => handleEnvironmentChange(value)}
                    >
                      {getEnvironnmentName(value)}
                    </Menu.Item>
                  ))}
                </Menu.Content>
              </Menu.Positioner>
            </Portal>
          </Menu.Root>
          <Activity mode={showPaasServerless ? "visible" : "hidden"}>
            <Button
              key="paasserverless"
              variant="ghost"
              onClick={() => {
                const paasserverless =
                  diagnostics?.extensions["paasserverless"];
                if (isExtensionInfo(paasserverless)) {
                  setExtension(paasserverless);
                  setSelectedTab("extensions");
                }
              }}
            >
              paasserverless
            </Button>
          </Activity>
          <Button
            key="websites"
            variant="ghost"
            onClick={() => {
              const websites = diagnostics?.extensions["websites"];
              if (isExtensionInfo(websites)) {
                setExtension(websites);
                setSelectedTab("extensions");
              }
            }}
          >
            websites
          </Button>
        </Flex>
        <ColorModeButton />
      </Flex>
      <Tabs.Root
        value={selectedTab}
        onValueChange={(details) => setSelectedTab(details.value)}
      >
        <Tabs.List>
          <Tabs.Trigger aria-controls="extensions-tab" value="extensions">
            Extensions
          </Tabs.Trigger>
          <Tabs.Trigger aria-controls="build-tab" value="build">
            Build Information
          </Tabs.Trigger>
          <Tabs.Trigger aria-controls="server-tab" value="server">
            Server Information
          </Tabs.Trigger>
        </Tabs.List>
      </Tabs.Root>
      <Activity mode={isLoading && !error ? "visible" : "hidden"}>
        <Box
          alignItems="center"
          display="flex"
          flex="1"
          justifyContent="center"
        >
          <Spinner aria-label="Loading..." />
        </Box>
      </Activity>
      <Activity mode={!isLoading && error ? "visible" : "hidden"}>
        <Box
          alignItems="center"
          display="flex"
          flex="1"
          justifyContent="center"
        >
          <Alert.Root status="error" maxW="md">
            <Alert.Indicator />
            <Alert.Title>Error</Alert.Title>
            <Alert.Description>{error?.message}</Alert.Description>
          </Alert.Root>
        </Box>
      </Activity>
      <Activity
        mode={
          !isLoading && !error && selectedTab === "extensions"
            ? "visible"
            : "hidden"
        }
      >
        {diagnostics?.extensions && (
          <Box id="extensions-tab" flex="1" overflowY="auto" role="tabpanel">
            <Flex flexDirection="row" gap="4" h="100%">
              <Extensions
                extensions={diagnostics.extensions}
                onLinkClick={handleLinkClick}
              />
              {extension && <Extension {...extension} />}
            </Flex>
          </Box>
        )}
      </Activity>
      <Activity
        mode={
          !isLoading && !error && selectedTab === "build" ? "visible" : "hidden"
        }
      >
        {diagnostics?.buildInfo && (
          <Box id="build-tab" flex="1" overflowY="auto" role="tabpanel">
            <BuildInfo {...diagnostics.buildInfo} />
          </Box>
        )}
      </Activity>
      <Activity
        mode={
          !isLoading && !error && selectedTab === "server"
            ? "visible"
            : "hidden"
        }
      >
        {diagnostics?.serverInfo && (
          <Box id="server-tab" flex="1" overflowY="auto" role="tabpanel">
            <ServerInfo {...diagnostics.serverInfo} />
          </Box>
        )}
      </Activity>
    </Flex>
  );
};

export default App;
