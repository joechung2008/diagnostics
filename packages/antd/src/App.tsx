import { Empty, Flex, Menu, Spin, Tabs } from "antd";
import { Activity, useCallback, useMemo, useState } from "react";
import useSWR from "swr";
import BuildInfo from "./BuildInfo";
import Extension from "./Extension";
import Extensions from "./Extensions";
import ServerInfo from "./ServerInfo";
import { isExtensionInfo, when } from "./utils";

import "./App.css";

const Environment = {
  Public: "https://hosting.portal.azure.net/api/diagnostics",
  Fairfax: "https://hosting.azureportal.usgovcloudapi.net/api/diagnostics",
  Mooncake: "https://hosting.azureportal.chinacloudapi.cn/api/diagnostics",
} as const;

type EnvironmentUrl = (typeof Environment)[keyof typeof Environment];

const items = [
  { key: "extensions", label: "Extensions" },
  { key: "build", label: "Build Information" },
  { key: "server", label: "Server Information" },
];

const App: React.FC = () => {
  const [extension, setExtension] = useState<ExtensionInfo>();
  const [environment, setEnvironment] = useState<EnvironmentUrl>(
    Environment.Public
  );
  const [selectedTab, setSelectedTab] = useState("extensions");

  const {
    data: diagnostics,
    error,
    isLoading,
  } = useSWR<Diagnostics>(environment, async (environment: string) => {
    const response = await fetch(environment);

    if (!response.ok) {
      throw new Error(`Error fetching diagnostics: ${response.statusText}`);
    }

    return response.json();
  });

  const environmentName = useMemo(() => {
    switch (environment) {
      case Environment.Public:
        return "Public Cloud";
      case Environment.Fairfax:
        return "Fairfax";
      case Environment.Mooncake:
        return "Mooncake";
      default:
        return "Select environment";
    }
  }, [environment]);

  const showPaasServerless = useMemo(
    () => isExtensionInfo(diagnostics?.extensions["paasserverless"]),
    [diagnostics?.extensions]
  );

  const environments = useMemo(
    () => [
      {
        key: "public",
        text: "Public Cloud",
        selected: environment === Environment.Public,
        onClick: () => {
          setEnvironment(Environment.Public);
          setExtension(undefined);
          setSelectedTab("extensions");
        },
      },
      {
        key: "fairfax",
        text: "Fairfax",
        selected: environment === Environment.Fairfax,
        onClick: () => {
          setEnvironment(Environment.Fairfax);
          setExtension(undefined);
          setSelectedTab("extensions");
        },
      },
      {
        key: "mooncake",
        text: "Mooncake",
        selected: environment === Environment.Mooncake,
        onClick: () => {
          setEnvironment(Environment.Mooncake);
          setExtension(undefined);
          setSelectedTab("extensions");
        },
      },
    ],
    [environment]
  );

  const menuItems = useMemo(
    () => [
      {
        key: "env-dropdown",
        label: environmentName,
        children: environments.map((env) => ({
          key: env.key,
          label: env.text,
          onClick: env.onClick,
        })),
      },
      ...when(showPaasServerless, {
        key: "paasserverless",
        label: "paasserverless",
        onClick: () => {
          const paasserverless = diagnostics?.extensions["paasserverless"];
          if (isExtensionInfo(paasserverless)) {
            setExtension(paasserverless);
            setSelectedTab("extensions");
          }
        },
      }),
      {
        key: "websites",
        label: "websites",
        onClick: () => {
          const websites = diagnostics?.extensions["websites"];
          if (isExtensionInfo(websites)) {
            setExtension(websites);
            setSelectedTab("extensions");
          }
        },
      },
    ],
    [environments, environmentName, showPaasServerless, diagnostics?.extensions]
  );

  const handleLinkClick = useCallback(
    (_?: React.MouseEvent, item?: KeyedNavLink) => {
      if (item) {
        const extension = diagnostics?.extensions[item.key];
        if (isExtensionInfo(extension)) {
          setExtension(extension);
        }
      }
    },
    [diagnostics?.extensions]
  );

  return (
    <div className="flexbox">
      <Menu mode="horizontal" items={menuItems} />
      <Tabs
        activeKey={selectedTab}
        className="tab-list"
        items={items}
        onChange={(key) => setSelectedTab(key)}
      />
      <Activity mode={isLoading ? "visible" : "hidden"}>
        <div className="tab-panel">
          <Flex className="centered">
            <Spin size="large" />
          </Flex>
        </div>
      </Activity>
      <Activity mode={error && !isLoading ? "visible" : "hidden"}>
        {error && (
          <div className="tab-panel">
            <Flex className="centered">
              <Empty
                description={`Error: ${error.message ?? "Unexpected error"}`}
              />
            </Flex>
          </div>
        )}
      </Activity>
      <Activity
        mode={
          !error && !isLoading && selectedTab === "extensions"
            ? "visible"
            : "hidden"
        }
      >
        {diagnostics?.extensions && (
          <div className="tab-panel">
            <Flex className="stack">
              <Extensions
                extensions={diagnostics.extensions}
                onLinkClick={handleLinkClick}
              />
              {extension && <Extension {...extension} />}
            </Flex>
          </div>
        )}
      </Activity>
      <Activity
        mode={
          !error && !isLoading && selectedTab === "build" ? "visible" : "hidden"
        }
      >
        {diagnostics?.buildInfo && (
          <div className="tab-panel">
            <BuildInfo {...diagnostics.buildInfo} />
          </div>
        )}
      </Activity>
      <Activity
        mode={
          !error && !isLoading && selectedTab === "server"
            ? "visible"
            : "hidden"
        }
      >
        {diagnostics?.buildInfo && (
          <div className="tab-panel">
            <ServerInfo {...diagnostics.serverInfo} />
          </div>
        )}
      </Activity>
    </div>
  );
};

export default App;
