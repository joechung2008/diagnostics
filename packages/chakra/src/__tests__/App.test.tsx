import { act, render, screen } from "@testing-library/react";
import useSWR from "swr";
import { beforeEach, describe, expect, it, vi } from "vitest";
import App from "../App";
import { TestWrapper } from "./test-utils";

// Mock useSWR
vi.mock("swr", () => ({
  default: vi.fn(),
}));

const mockUseSWR = vi.mocked(useSWR);

const mockDiagnosticsData = {
  buildInfo: { buildVersion: "1.2.3" },
  extensions: {
    testExtension: {
      extensionName: "Test Extension",
      config: { key1: "value1" },
      stageDefinition: { stage1: ["step1"] },
    },
    paasserverless: {
      extensionName: "PaaS Serverless",
      config: { enabled: "true" },
    },
    websites: {
      extensionName: "Websites",
      config: { version: "2.0" },
    },
  },
  serverInfo: {
    deploymentId: "deploy-123",
    extensionSync: { totalSyncAllCount: 42 },
    hostname: "server.example.com",
    nodeVersions: "v18.17.0",
    serverId: "server-456",
    uptime: 1234567890,
  },
};

describe("App", () => {
  beforeEach(() => {
    vi.clearAllMocks();
    mockUseSWR.mockReturnValue({
      data: mockDiagnosticsData,
      error: undefined,
      isLoading: false,
      mutate: vi.fn(),
      isValidating: false,
    });
  });

  it("should render loading state with spinner when isLoading is true", async () => {
    mockUseSWR.mockReturnValue({
      data: undefined,
      error: undefined,
      isLoading: true,
      mutate: vi.fn(),
      isValidating: false,
    });

    await act(async () => {
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      );
    });

    // When isLoading is true, the spinner should be present
    expect(screen.getByLabelText("Loading...")).toBeInTheDocument();
  });

  it("should render app with default environment", async () => {
    const { asFragment } = await act(async () =>
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      )
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display environment selector with default value", async () => {
    await act(async () => {
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      );
    });

    expect(
      screen.getByRole("button", { name: "Public Cloud" })
    ).toBeInTheDocument();
  });

  it("should render tabs for navigation", async () => {
    await act(async () => {
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      );
    });

    expect(screen.getByRole("tab", { name: "Extensions" })).toBeInTheDocument();
    expect(
      screen.getByRole("tab", { name: "Build Information" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("tab", { name: "Server Information" })
    ).toBeInTheDocument();
  });

  it("should display extensions tab by default", async () => {
    await act(async () => {
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      );
    });

    // Should show extensions content by default
    expect(screen.getByText("Test Extension")).toBeInTheDocument();
  });

  it("should not show paasserverless button when extension does not exist", async () => {
    mockUseSWR.mockReturnValue({
      data: {
        ...mockDiagnosticsData,
        extensions: {
          ...mockDiagnosticsData.extensions,
          paasserverless: undefined,
        },
      },
      error: undefined,
      isLoading: false,
      mutate: vi.fn(),
      isValidating: false,
    });

    await act(async () => {
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      );
    });

    expect(
      screen.queryByRole("button", { name: "paasserverless" })
    ).not.toBeInTheDocument();
  });

  it("should show websites button", async () => {
    await act(async () => {
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      );
    });

    expect(
      screen.getByRole("button", { name: "websites" })
    ).toBeInTheDocument();
  });

  it("should render color mode button", async () => {
    await act(async () => {
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      );
    });

    // Color mode button should be present
    expect(
      screen.getByRole("button", { name: /color mode/i })
    ).toBeInTheDocument();
  });
});
