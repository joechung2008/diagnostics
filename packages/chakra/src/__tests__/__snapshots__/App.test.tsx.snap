// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should render app with default environment 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-1l1i41u"
  >
    <div
      class="css-gg4vpm"
    >
      <div
        class="css-11ae5ff"
      >
        <button
          aria-haspopup="menu"
          class="chakra-button chakra-menu__trigger css-1ocnt5x"
          data-part="trigger"
          data-scope="menu"
          data-state="closed"
          data-uid="_r_2_"
          dir="ltr"
          id="menu:_r_2_:trigger"
          type="button"
        >
          Public Cloud
        </button>
        <button
          class="chakra-button css-1mi07d4"
          type="button"
        >
          paasserverless
        </button>
        <button
          class="chakra-button css-1mi07d4"
          type="button"
        >
          websites
        </button>
      </div>
      <button
        aria-label="Toggle color mode"
        class="chakra-button css-frz056"
        type="button"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="4"
          />
          <path
            d="M12 2v2"
          />
          <path
            d="M12 20v2"
          />
          <path
            d="m4.93 4.93 1.41 1.41"
          />
          <path
            d="m17.66 17.66 1.41 1.41"
          />
          <path
            d="M2 12h2"
          />
          <path
            d="M20 12h2"
          />
          <path
            d="m6.34 17.66-1.41 1.41"
          />
          <path
            d="m19.07 4.93-1.41 1.41"
          />
        </svg>
      </button>
    </div>
    <div
      class="chakra-tabs__root css-upaspd"
      data-orientation="horizontal"
      data-part="root"
      data-scope="tabs"
      dir="ltr"
      id="tabs:_r_3_"
    >
      <div
        aria-orientation="horizontal"
        class="chakra-tabs__list css-yamsmp"
        data-orientation="horizontal"
        data-part="list"
        data-scope="tabs"
        dir="ltr"
        id="tabs:_r_3_:list"
        role="tablist"
      >
        <button
          aria-controls="extensions-tab"
          aria-selected="true"
          class="chakra-tabs__trigger css-19t98pl"
          data-focus=""
          data-orientation="horizontal"
          data-ownedby="tabs:_r_3_:list"
          data-part="trigger"
          data-scope="tabs"
          data-selected=""
          data-value="extensions"
          dir="ltr"
          id="tabs:_r_3_:trigger-extensions"
          role="tab"
          tabindex="0"
          type="button"
        >
          Extensions
        </button>
        <button
          aria-controls="build-tab"
          aria-selected="false"
          class="chakra-tabs__trigger css-19t98pl"
          data-orientation="horizontal"
          data-ownedby="tabs:_r_3_:list"
          data-part="trigger"
          data-scope="tabs"
          data-value="build"
          dir="ltr"
          id="tabs:_r_3_:trigger-build"
          role="tab"
          tabindex="-1"
          type="button"
        >
          Build Information
        </button>
        <button
          aria-controls="server-tab"
          aria-selected="false"
          class="chakra-tabs__trigger css-19t98pl"
          data-orientation="horizontal"
          data-ownedby="tabs:_r_3_:list"
          data-part="trigger"
          data-scope="tabs"
          data-value="server"
          dir="ltr"
          id="tabs:_r_3_:trigger-server"
          role="tab"
          tabindex="-1"
          type="button"
        >
          Server Information
        </button>
      </div>
    </div>
    <div
      class="css-1dwgpej"
      style="display: none;"
    >
      <span
        aria-label="Loading..."
        class="chakra-spinner css-11t4vt7"
      />
    </div>
    <div
      class="css-1dwgpej"
      style="display: none;"
    >
      <div
        class="chakra-alert__root css-1mc1tb4"
      >
        <span
          class="css-13gnd3a"
        >
          <svg
            class="css-126y6n1"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2C17.5228 2 22 6.47715 22 12C22 17.5228 17.5228 22 12 22ZM12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20ZM11 15H13V17H11V15ZM11 7H13V13H11V7Z"
            />
          </svg>
        </span>
        <div
          class="chakra-alert__title css-1kctzi1"
        >
          Error
        </div>
        <div
          class="chakra-alert__description css-t6kp1r"
        />
      </div>
    </div>
    <div
      class="css-177dxab"
      id="extensions-tab"
      role="tabpanel"
    >
      <div
        class="css-e3v19i"
      >
        <nav
          aria-label="Extensions"
          class="css-rqcg1t"
        >
          <button
            class="chakra-button css-1tlpiy1"
            type="button"
          >
            PaaS Serverless
          </button>
          <button
            class="chakra-button css-1tlpiy1"
            type="button"
          >
            Test Extension
          </button>
          <button
            class="chakra-button css-1tlpiy1"
            type="button"
          >
            Websites
          </button>
        </nav>
      </div>
    </div>
    <div
      class="css-177dxab"
      id="build-tab"
      role="tabpanel"
      style="display: none;"
    >
      <table
        class="chakra-table__root css-a2tetu"
      >
        <thead
          class="chakra-table__header"
        >
          <tr
            class="chakra-table__row css-11ot24d"
          >
            <th
              class="chakra-table__columnHeader css-kim7sg"
            >
              Name
            </th>
            <th
              class="chakra-table__columnHeader css-kim7sg"
            >
              Value
            </th>
          </tr>
        </thead>
        <tbody
          class="chakra-table__body"
        >
          <tr
            class="chakra-table__row css-11ot24d"
          >
            <td
              class="chakra-table__cell css-1luypx0"
            >
              Build Version
            </td>
            <td
              class="chakra-table__cell css-1luypx0"
            >
              1.2.3
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="css-177dxab"
      id="server-tab"
      role="tabpanel"
      style="display: none;"
    >
      <table
        class="chakra-table__root css-a2tetu"
      >
        <thead
          class="chakra-table__header"
        >
          <tr
            class="chakra-table__row css-11ot24d"
          >
            <th
              class="chakra-table__columnHeader css-kim7sg"
            >
              Name
            </th>
            <th
              class="chakra-table__columnHeader css-kim7sg"
            >
              Value
            </th>
          </tr>
        </thead>
        <tbody
          class="chakra-table__body"
        >
          <tr
            class="chakra-table__row css-11ot24d"
          >
            <td
              class="chakra-table__cell css-1luypx0"
            >
              Hostname
            </td>
            <td
              class="chakra-table__cell css-1luypx0"
            >
              server.example.com
            </td>
          </tr>
          <tr
            class="chakra-table__row css-11ot24d"
          >
            <td
              class="chakra-table__cell css-1luypx0"
            >
              Uptime
            </td>
            <td
              class="chakra-table__cell css-1luypx0"
            >
              1234567890
            </td>
          </tr>
          <tr
            class="chakra-table__row css-11ot24d"
          >
            <td
              class="chakra-table__cell css-1luypx0"
            >
              Server ID
            </td>
            <td
              class="chakra-table__cell css-1luypx0"
            >
              server-456
            </td>
          </tr>
          <tr
            class="chakra-table__row css-11ot24d"
          >
            <td
              class="chakra-table__cell css-1luypx0"
            >
              Deployment ID
            </td>
            <td
              class="chakra-table__cell css-1luypx0"
            >
              deploy-123
            </td>
          </tr>
          <tr
            class="chakra-table__row css-11ot24d"
          >
            <td
              class="chakra-table__cell css-1luypx0"
            >
              Node Versions
            </td>
            <td
              class="chakra-table__cell css-1luypx0"
            >
              v18.17.0
            </td>
          </tr>
          <tr
            class="chakra-table__row css-11ot24d"
          >
            <td
              class="chakra-table__cell css-1luypx0"
            >
              Extension Sync | Total Sync All Count
            </td>
            <td
              class="chakra-table__cell css-1luypx0"
            >
              42
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
